File: src/main/java/org/hildan/bots/riseoflords/network/parsers/GoldImageOCR.java
package org.hildan.bots.riseoflords.network.parsers;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import javax.imageio.ImageIO;

import org.hildan.bots.riseoflords.util.Log;

public class GoldImageOCR {

    private static final String DIGITS_DIR = "/img";

    private static final BufferedImage[] DIGITS = { loadImage(DIGITS_DIR + "/0.png"), loadImage(DIGITS_DIR + "/1.png"),
        loadImage(DIGITS_DIR + "/2.png"), loadImage(DIGITS_DIR + "/3.png"), loadImage(DIGITS_DIR + "/4.png"),
        loadImage(DIGITS_DIR + "/5.png"), loadImage(DIGITS_DIR + "/6.png"), loadImage(DIGITS_DIR + "/7.png"),
        loadImage(DIGITS_DIR + "/8.png"), loadImage(DIGITS_DIR + "/9.png") };

    private static final BufferedImage DOT = loadImage(DIGITS_DIR + "/dot.png");

    public static int readAmount(BufferedImage img) {
        assert img.getWidth() == 70 : "image width is not 70";
        assert img.getHeight() == 8 : "image height is not 8";
        final List<BufferedImage> digits = getDigitsImages(img);
        final StringBuilder sb = new StringBuilder();
        for (final BufferedImage digit : digits) {
            sb.append(getDigit(digit));
        }
        try {
            final String str = sb.toString();
            if (str.equals("")) {
                System.err.println("OCR failed to recognize anything");
                return -1;
            }
            return Integer.valueOf(str.replace(".", ""));
        } catch (final NumberFormatException e) {
            System.err.println("bad OCR result");
            return -1;
        }
    }

    private static BufferedImage loadImage(String filename) {
        try {
            final BufferedImage img = ImageIO.read(GoldImageOCR.class.getResourceAsStream(filename));
            return img;
        } catch (final IOException e) {
            throw new RuntimeException("image couldn't be loaded");
        }
    }

    private static List<Integer> getEmptyColumns(BufferedImage img) {
        final List<Integer> emptyCols = new ArrayList<>();
        col_loop: for (int i = 0; i < img.getWidth(); i++) {
            for (int j = 0; j < img.getHeight(); j++) {
                final int rgb = img.getRGB(i, j);
                if (!(rgb >>> 24 == 0)) {
                    continue col_loop;
                }
            }
            emptyCols.add(i);
        }
        return emptyCols;
    }

    private static List<Integer[]> getDigitsBounds(BufferedImage img) {
        final List<Integer> emptyCols = getEmptyColumns(img);
        final List<Integer[]> digitsBounds = new ArrayList<>();
        Integer start = null;
        Integer end = null;
        for (int i = 0; i < img.getWidth(); i++) {
            if (emptyCols.contains(i)) {
                if (start != null && end != null) {
                    digitsBounds.add(new Integer[] { start, end });
                }
                start = null;
                end = null;
                continue;
            }
            if (start == null) {
                start = i;
            }
            end = i;
        }
        return digitsBounds;
    }

    private static List<BufferedImage> getDigitsImages(BufferedImage img) {
        final List<Integer[]> digitsBounds = getDigitsBounds(img);
        final List<BufferedImage> digitsImages = new ArrayList<>();
        for (final Integer[] bounds : digitsBounds) {
            final BufferedImage digitImg = new BufferedImage(bounds[1] - bounds[0] + 1, img.getHeight(),
                    BufferedImage.TYPE_INT_ARGB);
            for (int i = 0; i < digitImg.getWidth(); i++) {
                for (int j = 0; j < digitImg.getHeight(); j++) {
                    digitImg.setRGB(i, j, img.getRGB(i + bounds[0], j));
                }
            }
            digitsImages.add(digitImg);
        }
        return digitsImages;
    }

    private static int[] getARGB(int argb) {
        final int[] res = new int[4];
        res[0] = argb >>> 24 & 0xFF;
            res[1] = argb >>> 16 & 0xFF;
            res[2] = argb >>> 8 & 0xFF;
            res[3] = argb & 0xFF;
            return res;
    }

    private static boolean areSimilar(int recoPixel, int refPixel) {
        if (recoPixel == refPixel) {
            return true;
        }
        final int[] recoARGB = getARGB(recoPixel);
        final int[] refARGB = getARGB(refPixel);
        if (recoARGB[0] == refARGB[0]) {
            return true;
        }
        return false;
    }

    private static boolean areSimilar(BufferedImage recoDigit, BufferedImage refDigit) {
        // check dimensions
        if (recoDigit.getHeight() != refDigit.getHeight()) {
            return false;
        }
        if (recoDigit.getWidth() != refDigit.getWidth()) {
            return false;
        }
        // check pixels
        for (int i = 0; i < recoDigit.getWidth(); i++) {
            for (int j = 0; j < recoDigit.getHeight(); j++) {
                final int recoPixel = recoDigit.getRGB(i, j);
                final int refPixel = refDigit.getRGB(i, j);
                if (!areSimilar(recoPixel, refPixel)) {
                    return false;
                }
            }
        }
        return true;
    }

    private static String getDigit(BufferedImage digitImg) {
        for (int i = 0; i < DIGITS.length; i++) {
            if (areSimilar(digitImg, DIGITS[i])) {
                return String.valueOf(i);
            }
        }
        if (areSimilar(digitImg, DOT)) {
            return ".";
        }
        Log.get().e("OCR", "unrecognized digit");
        return " ";
    }

}

File: src/main/kotlin/org/hildan/bots/riseoflords/network/parsers/GoldImageOCR.kt
package org.hildan.bots.riseoflords.network.parsers

import org.slf4j.LoggerFactory
import java.awt.image.BufferedImage
import java.io.IOException
import java.util.*
import javax.imageio.ImageIO

internal object GoldImageOCR {

    private val logger = LoggerFactory.getLogger(GoldImageOCR::class.java)

    private const val DIGITS_DIR = "/img"

    private val DIGITS = arrayOf(
        loadInternalImageResource("$DIGITS_DIR/0.png"),
        loadInternalImageResource("$DIGITS_DIR/1.png"),
        loadInternalImageResource("$DIGITS_DIR/2.png"),
        loadInternalImageResource("$DIGITS_DIR/3.png"),
        loadInternalImageResource("$DIGITS_DIR/4.png"),
        loadInternalImageResource("$DIGITS_DIR/5.png"),
        loadInternalImageResource("$DIGITS_DIR/6.png"),
        loadInternalImageResource("$DIGITS_DIR/7.png"),
        loadInternalImageResource("$DIGITS_DIR/8.png"),
        loadInternalImageResource("$DIGITS_DIR/9.png")
    )

    private val DOT = loadInternalImageResource("$DIGITS_DIR/dot.png")

    fun readAmount(img: BufferedImage): Int {
        assert(img.width == 70) { "image width is not 70" }
        assert(img.height == 8) { "image height is not 8" }
        val digits = splitIntoDigits(img)
        val sb = StringBuilder()
        for (digit in digits) {
            sb.append(recognizeDigitOrDot(digit))
        }
        val amountAsText = sb.toString()
        return try {
            if (amountAsText == "") {
                logger.error("OCR failed to recognize anything")
                return -1
            }
            Integer.valueOf(amountAsText.replace(".", ""))
        } catch (e: NumberFormatException) {
            logger.error("Bad OCR result, cannot convert '{}' to a gold amount", amountAsText)
            -1
        }
    }

    private fun loadInternalImageResource(filename: String): BufferedImage = try {
        ImageIO.read(GoldImageOCR::class.java.getResourceAsStream(filename))
    } catch (e: IOException) {
        throw RuntimeException("Internal image $filename couldn't be loaded")
    }

    private fun getEmptyColumns(img: BufferedImage): List<Int> {
        val emptyCols: MutableList<Int> = ArrayList()
        col_loop@ for (i in 0 until img.width) {
            for (j in 0 until img.height) {
                val rgb = img.getRGB(i, j)
                if (rgb ushr 24 != 0) {
                    continue@col_loop
                }
            }
            emptyCols.add(i)
        }
        return emptyCols
    }

    private fun getDigitsBounds(img: BufferedImage): List<Array<Int?>> {
        val emptyCols = getEmptyColumns(img)
        val digitsBounds: MutableList<Array<Int?>> = ArrayList()
        var start: Int? = null
        var end: Int? = null
        for (i in 0 until img.width) {
            if (emptyCols.contains(i)) {
                if (start != null) {
                    digitsBounds.add(arrayOf(start, end))
                }
                start = null
                end = null
                continue
            }
            if (start == null) {
                start = i
            }
            end = i
        }
        return digitsBounds
    }

    private fun splitIntoDigits(img: BufferedImage): List<BufferedImage> {
        val digitsBounds = getDigitsBounds(img)
        val digitsImages: MutableList<BufferedImage> = ArrayList()
        for (bounds in digitsBounds) {
            val digitImg = BufferedImage(bounds[1]!! - bounds[0]!! + 1, img.height, BufferedImage.TYPE_INT_ARGB)
            for (i in 0 until digitImg.width) {
                for (j in 0 until digitImg.height) {
                    digitImg.setRGB(i, j, img.getRGB(i + bounds[0]!!, j))
                }
            }
            digitsImages.add(digitImg)
        }
        return digitsImages
    }

    private fun getARGB(argb: Int): IntArray {
        val res = IntArray(4)
        res[0] = argb ushr 24 and 0xFF
        res[1] = argb ushr 16 and 0xFF
        res[2] = argb ushr 8 and 0xFF
        res[3] = argb and 0xFF
        return res
    }

    private fun areSimilar(recoPixel: Int, refPixel: Int): Boolean {
        if (recoPixel == refPixel) {
            return true
        }
        val recoARGB = getARGB(recoPixel)
        val refARGB = getARGB(refPixel)
        return recoARGB[0] == refARGB[0]
    }

    private fun areSimilar(candidate: BufferedImage, reference: BufferedImage): Boolean {
        // check dimensions
        if (candidate.height != reference.height) {
            return false
        }
        if (candidate.width != reference.width) {
            return false
        }
        // check pixels
        for (i in 0 until candidate.width) {
            for (j in 0 until candidate.height) {
                val recoPixel = candidate.getRGB(i, j)
                val refPixel = reference.getRGB(i, j)
                if (!areSimilar(recoPixel, refPixel)) {
                    return false
                }
            }
        }
        return true
    }

    private fun recognizeDigitOrDot(digitImg: BufferedImage): String {
        for (i in DIGITS.indices) {
            if (areSimilar(digitImg, DIGITS[i])) {
                return i.toString()
            }
        }
        if (areSimilar(digitImg, DOT)) {
            return "."
        }
        logger.error("Unrecognized digit in gold image")
        return " "
    }
}